src/Common/HTTPConnectionPool.cpp

```diff
@@ -0,0 +1,865 @@
+// - soft limit, warn limit, store limit
+// - `Session::reconnect()` uses the pool as well
+// - comprehensive sensors
+// - session is reused according its inner state, automatically
+
+
+template <class Session>
+class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConnectionPool<Session>>, public IExtendedPool
+{
+private:
+    friend class HTTPConnectionPools;
+
+    using WeakPtr = std::weak_ptr<EndpointConnectionPool<Session>>;
+
+    class PooledConnection : public Session
+    {
+    public:
+        using Ptr = std::shared_ptr<PooledConnection>;
+
+        void reconnect() override
+        {
+            ProfileEvents::increment(metrics.reset);
+            Session::close();
+
+            if (auto lock = pool.lock())
+            {
+                auto timeouts = getTimeouts(*this);
+                auto new_connection = lock->getConnection(timeouts);
+                Session::assign(*new_connection);
+            }
+            else
+            {
+                auto timer = CurrentThread::getProfileEvents().timer(metrics.elapsed_microseconds);
+                Session::reconnect();
+                ProfileEvents::increment(metrics.created);
+            }
+        }
+
+        String getTarget() const
+        {
+            if (!Session::getProxyConfig().host.empty())
+                return fmt::format("{} over proxy {}", Session::getHost(), Session::getProxyConfig().host);
+            return Session::getHost();
+        }
+
+        void flushRequest() override
+        {
+            if (bool(request_stream))
+            {
+                request_stream->flush();
+
+                if (auto * fixed_steam = dynamic_cast<Poco::Net::HTTPFixedLengthOutputStream *>(request_stream))
+                {
+                    request_stream_completed = fixed_steam->isComplete();
+                }
+                else if (auto * chunked_steam = dynamic_cast<Poco::Net::HTTPChunkedOutputStream *>(request_stream))
+                {
+                    chunked_steam->rdbuf()->close();
+                    request_stream_completed = chunked_steam->isComplete();
+                }
+                else if (auto * http_stream = dynamic_cast<Poco::Net::HTTPOutputStream *>(request_stream))
+                {
+                    request_stream_completed = http_stream->isComplete();
+                }
+                else
+                {
+                    request_stream_completed = false;
+                }
+            }
+            request_stream = nullptr;
+
+            Session::flushRequest();
+        }
+
+        std::ostream & sendRequest(Poco::Net::HTTPRequest & request) override
+        {
+            std::ostream & result = Session::sendRequest(request);
+            result.exceptions(std::ios::badbit);
+
+            request_stream = &result;
+            request_stream_completed = false;
+
+            response_stream = nullptr;
+            response_stream_completed = true;
```

Comment:
Why `true`? Won't this make us incorrectly reuse the session if we do `sendRequest()`, then `flushRequest()`, then destroy the pooled connection?

Reply:
I remember that I intentionally made this scenario preserve the session. 
But you are right, that scenario has to reset the session. I have added a test for it.